futaba/cogs/moderation/cleanup.py

```diff
@@ -296,3 +318,186 @@ def check(message):
+            "This may take a while..."
+        )
+        status_message = await ctx.send(embed=embed)
+
+        # Deletion condition function
+        def check(message):
+            if flag.value:
+                # If this is set, then the deletion has been cancelled.
+                raise ValueError
+
+            return message.author == user
+
+        # Run deletions
+        try:
+            deleted = await self.purge_all_messages(guild, user, check, status_message)
+        except ValueError:
+            elapsed = datetime.now() - start
+            embed = discord.Embed(colour=discord.Colour.red())
+            embed.title = "Complete user message purge cancelled"
+            embed.description = (
+                f"Deletion was cancelled. Spent {fancy_timedelta(elapsed)} in deletion."
+            )
+            raise CommandFailed(embed=embed)
+
+        # Notify that deletions are finished
+        elapsed = datetime.now() - start
+        embed = discord.Embed(colour=discord.Colour.dark_teal())
+        embed.title = "Deletion complete"
+        embed.description = (
+            f"Deleted all {len(deleted)} messages sent by {user.mention}. "
+            f"(took {fancy_timedelta(elapsed)})"
+        )
+        await ctx.send(embed=embed)
+
+        # Remove code from map
+        del self.delete_codes[code]
+
+    async def purge_all_messages(self, guild, user, check, status_message):
+        """
+        Implementation function for actually removing all messages by a user.
+        Don't run this directly!
+        """
+
+        # Task for deleting all messages within a channel
+        async def purge_channel(channel):
+            logger.debug(
+                "Starting purge of user '%s' (%d) in channel '%s' (%d)",
+                user.name,
+                user.id,
+                channel.name,
+                channel.id,
+            )
+
+            deleted = await channel.purge(limit=None, check=check, bulk=True)
+
+            logger.debug(
+                "Finished purge of user '%s' (%d) in channel '%s' (%d)",
+                user.name,
+                user.id,
+                channel.name,
+                channel.id,
+            )
+
+            # Update status message
+            embed = discord.Embed(colour=discord.Colour.dark_teal())
+            embed.title = "Complete user message purge"
+            embed.description = (
+                f"Commencing deletion of **all** messages sent by {user.mention} across all channels.\n"
+                f"This may take a while... (finished channel {channel.mention})"
+            )
+            await status_message.edit(embed=embed)
+
+            return deleted
+
+        # Create individual tasks
+        tasks = []
+
+        for channel in guild.text_channels:
+            tasks.append(purge_channel(channel))
+
+        # Run them in parallel
+        all_deleted = await asyncio.gather(*tasks)
```

Comment:
My only question is if we run this in parallel would this affect other tasks (sending journal events to log channels, users running commands). Since wouldn't this be hitting discords API hard if a user has a lot of messages

Reply:
That's true. I was going to suggest adding a delay to `check()` but it's not async. 

Though this command should be run seldom enough that this won't be a big issue with performance I think

Reply:
Yeah good point. I'm fine with it being added

Reply:
discord.py takes care of global rate-limiting https://github.com/Rapptz/discord.py/blob/b77af199392209017e59eff366b46354fa210f64/discord/http.py#L223

Reply:
@raylu It unfortunately doesn't sleep when purging a text channel though, so it uses up the available events all at once. So it'll delete ~120 messages in a spurt, then wait a minute, then delete another ~120, clogging up the whole system (number of messages depending on when other queued requests are filled).

Reply:
It is still running other commands when it can, so for now it should be okay it just takes a bit for it to do so. Plus we do have other events that will wait for down time like the journal logging